xcm/xcm-builder/src/fee_handling.rs

```diff
@@ -0,0 +1,58 @@
+// Copyright Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use core::marker::PhantomData;
+use frame_support::traits::{Contains, Get};
+use xcm::prelude::*;
+use xcm_executor::traits::{FeeManager, FeeReason, TransactAsset};
+
+/// A `FeeManager` implementation that simply deposits the fees handled into a specific on-chain
+/// `ReceiverAccount`.
+///
+/// It reuses the `AssetTransactor` configured on the XCM executor to deposit fee assets, and also
+/// permits specifying `WaivedLocations` for locations that are privileged to not pay for fees. If
+/// the `AssetTransactor` returns an error while calling `deposit_asset`, then a warning will be
+/// logged.
+pub struct XcmFeesToAccount<XcmConfig, WaivedLocations, AccountId, ReceiverAccount>(
+	PhantomData<(XcmConfig, WaivedLocations, AccountId, ReceiverAccount)>,
+);
+impl<
+		XcmConfig: xcm_executor::Config,
+		WaivedLocations: Contains<MultiLocation>,
+		AccountId: Clone + Into<[u8; 32]>,
+		ReceiverAccount: Get<Option<AccountId>>,
+	> FeeManager for XcmFeesToAccount<XcmConfig, WaivedLocations, AccountId, ReceiverAccount>
+{
+	fn is_waived(origin: Option<&MultiLocation>, _: FeeReason) -> bool {
+		let Some(loc) = origin else { return false };
+		WaivedLocations::contains(loc)
+	}
+
+	fn handle_fee(fees: MultiAssets, context: Option<&XcmContext>) {
+		if let Some(receiver) = ReceiverAccount::get() {
+			let dest = AccountId32 { network: None, id: receiver.into() }.into();
+			for asset in fees.into_inner() {
+				if let Err(e) = XcmConfig::AssetTransactor::deposit_asset(&asset, &dest, context) {
```

Comment:
How about using fungibles trait?
I think we need only to convert `MultiAssets` with the same `MatchesFungibles` to `asset_id` and `amount`.

Reply:
just out of curiosity, what fungibles trait?
AssetTransactor does all conversions (and it doesnt matter if Fungibles or NonFungibles)

Reply:
[This one for example](https://github.com/paritytech/substrate/blob/65c71a059e7f0e48b6c63009e06de907d4fe12c9/frame/support/src/traits/tokens/fungibles/regular.rs#L474), implemented by assets or balances pallets

Reply:
yeah, but you can aggregate those fungible/s impls as well. 
`AssetTransactor` also converts the `dest`, it might be not what the user of `XcmFeesToAccount` wants/expects.
also, using fungibles impls directly does not require a context, which probably not meant for `handle_fee`.
current solution looks good to me as well, just suggestion.

Reply:
This is 100% a bad idea. The fungibles traits are in `frame_support` because they are a concept in FRAME, and we should be introducing less coupling between FRAME and XCM, not more.

Reply:
`xcm-builder` depends from `frame_support `. The `CurrencyAdapter` which part of `xcm-builder` depends from fungibles.
`xcm` protocol itself should not depend from `frame`, my humble opinion. And its more realistic to make `xcm-executor` free from frame.